README.md
<p align="center"><img src="https://i.imgur.com/yB3GZ9j.png"></p>

<p align="center">RelativeScale is a simple custom sizing system for flutter widgets to achieve the same physical sizes across different devices.</p>

<p align="center">
<a href="https://pub.dev/packages/relative_scale" target="_blank"><img src="https://img.shields.io/pub/v/relative_scale" alt="Pub Version" /></a>
<a href="https://github.com/xamantra/relative_scale/actions?query=workflow%3ACI" target="_blank"><img src="https://github.com/xamantra/relative_scale/workflows/CI/badge.svg?event=push" alt="CI" /></a>
<a href="https://codecov.io/gh/xamantra/relative_scale"><img src="https://codecov.io/gh/xamantra/relative_scale/branch/master/graph/badge.svg" /></a>
<a href="https://pub.dev/packages/relative_scale/score" target="_blank"><img src="https://img.shields.io/badge/dynamic/json?color=blue&label=likes&query=likes&url=http://www.pubscore.gq/likes?package=relative_scale" alt="likes" /></a>
<a href="https://pub.dev/packages/relative_scale/score" target="_blank"><img src="https://img.shields.io/badge/dynamic/json?color=green&label=health&query=pub_points&url=http://www.pubscore.gq/pub-points?package=relative_scale" alt="health" /></a>
<a href="https://pub.dev/packages/relative_scale/score" target="_blank"><img src="https://img.shields.io/badge/dynamic/json?color=teal&label=popularity&query=popularity&url=http://www.pubscore.gq/popularity?package=relative_scale" alt="popularity" /></a>
<a href="https://github.com/xamantra/relative_scale/stargazers" target="_blank"><img src="https://img.shields.io/github/stars/xamantra/relative_scale" alt="GitHub stars" /></a>
<a href="https://github.com/tenhobi/effective_dart" target="_blank"><img src="https://img.shields.io/badge/style-effective_dart-54C5F8.svg" alt="style: effective dart" /></a>
<a href="https://github.com/xamantra/relative_scale/blob/master/LICENSE" target="_blank"><img src="https://img.shields.io/github/license/xamantra/relative_scale" alt="GitHub license" /></a>
<a href="https://github.com/xamantra/relative_scale/commits/master" target="_blank"><img src="https://img.shields.io/github/last-commit/xamantra/relative_scale" alt="GitHub last commit" /></a>
</p>

---

## Usage
It is VERY easy to use.

- `sy(value)` and `sx(value)`
- Example:
  - `sy(10`) - size relative to screen height.
  - `sx(10`) - size relative to screen width.

- `RelativeBuilder` - A builder widget that provides the scaling methods `sy` and `sx`.
- `RelativeScale` (`deprecated`) - A StatefulWidget mixin that provides the scaling methods `sy` and `sx`. It is highly recommended to use `RelativeBuilder`.

<hr>

## Example
`RelativeBuilder` example.

```dart
class Example extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return RelativeBuilder(
      builder: (context, height, width, sy, sx) {
        return Scaffold(
          appBar: AppBar(
            title: Text(
              'App Bar Title',
              style: TextStyle(fontSize: sy(13)),
            ),
          ),
          body: Container(
            height: height,
            width: width,
            child: Text(
              'Body Text',
              style: TextStyle(fontSize: sy(11)),
            ),
          ),
        );
      },
    );
  }
}
```

<hr>

## Preview

The base screen size for these widgets is `480 x 800`. The source code for this preview app is [here](https://github.com/xamantra/flutter_relative_scale_example).

### Scaled with `RelativeScale`:

![Scaled](./preview/scaled.png)

Now, look at these scaled widgets with _RelativeScale_. There is a difference, yeah, but that's because of the system scaled sizes like the AppBar (look at the appbar's height :) ). Now let's forget about that and focus on the texts and the rectangle containers. They are the same sizes.

### Normal scaling from a flutter:

  ![Unscaled](./preview/unscaled.png)

  Now, for unscaled sizes, no _RelativeScale_ at all. Well, that's quite obvious :). Look at the texts on the last image, they are very small comparing to the first image. And the rectangle containers, very big difference.


The full example is in the [Example](https://pub.dev/packages/relative_scale#-example-tab-) section.

<hr>

## Scaling Notes

For instance, you have a container widget like this:

```Dart
Container(
    height: 300,
    width: 500,
)
```

and you implemented RelativeScale:

```Dart
Container(
    height: sy(300),
    width: sx(500),
)
```

they will not be the same size anymore, using relative scaler will make your sizes a bit bigger. But the hard work will payoff after adjusting your sizes because your app will now have the same widget sizes in every devices.

Please note that these scaler methods are relative to screen size. So basically in this case `sy(50)` and `sx(50)` is NOT the same size.

Also, another thing to note is that if you use `sy` for height and `sx` for width (or vice-versa), you'll get widgets with the same ratio (not size) which is still useful. The _Scaled_ preview image above uses only `sy`, and containers and text have the same size across different screens.

#### If you want to make a perfect _Square_ container, DON'T do this:

```Dart
Container(
    height: sy(300),
    width: sx(300),
)
// Yeah they are the same value "300", but they are not the same unit 'cause you used "sx" on the width.
```

#### DO this instead:

```Dart
Container(
    height: sy(300), // or sx(value)
    width: sy(300), // or sx(value)
)
```

## Portrait & Landscape
The library doesn't support orientation yet. If you designed your app for portrait and the user rotates the app to landscape view, the sizes will be messed up. So if you want to use this library, it's highly recommended you lock the rotation. Or if your app should adapt to orientation, **DO NOT** use this library, there are a lot of alternatives.

CHANGELOG.md
## 2.0.0
- Null-safety.
- Flutter 2.0.0.

## 1.1.2
- Formatting with `dartfmt -w lib`.

<hr>

## 1.1.1
- Deprecated `RelativeScale`.
- Updated readme.

<hr>

## 1.1.0

- hosted the logo in imgur instead of in root directory

<hr>

## 1.0.9

- Readme update.
- added internal tests.
- added logo.
- remove unused files in example/ directory.

<hr>

## 1.0.8

- Readme patch. Added badges and fixed the grammars.

<hr>

## 1.0.7

- **NEW:** `RelativeBuilder` which can be used for stateless widgets, you can also use this in stateful widgets. Head to [example](https://pub.dev/packages/relative_scale#-example-tab-) tab.

<hr>

## 1.0.6

- updated readme and package description.

<hr>

## 1.0.5

- Added link to the preview app.

<hr>

## 1.0.4

- Added documentation and previews on readme and updated example.dart

<hr>

## 1.0.3

- Rename the mixin `RelativeScale` to `RelativeScale` to follow package name.

<hr>

## 1.0.2

- Updated library description documentation.

<hr>

## 1.0.1

- Added example.dart

<hr>

## 1.0.0

- Initial version.


test/main_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'widget.relative_builder.dart';
import 'widget.relative_scale.dart';

void main() {
  testWidgets('RelativeScale mixin test', (tester) async {
    await tester.pumpWidget(RelativeScaleWidget());
    await tester.pumpAndSettle();

    expect(tester.takeException(), null);
  });

  testWidgets('RelativeBuilder widget test', (tester) async {
    await tester.pumpWidget(RelativeBuilderWidget());
    await tester.pumpAndSettle();

    expect(tester.takeException(), null);
  });
}


test/widget.relative_scale.dart
import 'package:flutter/material.dart';
import 'package:relative_scale/relative_scale.dart';

class RelativeScaleWidget extends StatelessWidget {
  const RelativeScaleWidget({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: TestWidget(),
    );
  }
}

class TestWidget extends StatefulWidget {
  TestWidget({Key? key}) : super(key: key);

  @override
  _TestWidgetState createState() => _TestWidgetState();
}

// ignore: deprecated_member_use_from_same_package
class _TestWidgetState extends State<TestWidget> with RelativeScale {
  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    initRelativeScaler(context);
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: screenHeight,
      width: screenWidth,
      child: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            SizedBox(
              height: sy(30),
              width: sy(30),
            ),
            SizedBox(
              height: sx(30),
              width: sx(30),
            ),
          ],
        ),
      ),
    );
  }
}


test/widget.relative_builder.dart
import 'package:flutter/material.dart';
import 'package:relative_scale/relative_scale.dart';

class RelativeBuilderWidget extends StatelessWidget {
  const RelativeBuilderWidget({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: TestWidget(),
    );
  }
}

class TestWidget extends StatelessWidget {
  const TestWidget({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return RelativeBuilder(
      builder: (context, height, width, sy, sx) {
        return Container(
          height: height,
          width: width,
          child: Center(
            child: Column(
              mainAxisSize: MainAxisSize.min,
              children: [
                SizedBox(
                  height: sy(30),
                  width: sy(30),
                ),
                SizedBox(
                  height: sx(30),
                  width: sx(30),
                ),
              ],
            ),
          ),
        );
      },
    );
  }
}


example/stateful.dart
import 'package:flutter/material.dart';
import 'package:relative_scale/relative_scale.dart';

class Example extends StatefulWidget {
  Example({Key key}) : super(key: key);

  @override
  _ExampleState createState() => _ExampleState();
}

// RelativeScaler is a mixin
class _ExampleState extends State<Example> with RelativeScale {
  @override
  Widget build(BuildContext context) {
    // this is very important, you should always call this whenever you implement RelativeScaler on a widget.
    initRelativeScaler(context);

    /* You'll see "sy" and "sx" below: */
    /* which means "sy" = size relative to height of the screen. */
    /* which means "sx" = size relative to width of the screen. */

    return Container(
      height: screenHeight,
      width: screenWidth,
      child: SingleChildScrollView(
        child: Column(
          children: <Widget>[
            Container(
              height: sy(160),
              width: sy(240),
              margin: EdgeInsets.all(sy(12)),
              color: Colors.redAccent,
              child: Center(
                child: Text(
                  "Scaled Text 1",
                  style: TextStyle(
                    fontSize: sy(24),
                    color: Colors.white,
                  ),
                ),
              ),
            ),
            Container(
              height: sy(80),
              width: sy(280),
              margin: EdgeInsets.all(sy(12)),
              color: Colors.green,
              child: Center(
                child: Text(
                  "Scaled Text 2",
                  style: TextStyle(
                    fontSize: sy(24),
                    color: Colors.white,
                  ),
                ),
              ),
            ),
            Container(
              height: sy(120),
              width: sy(240),
              margin: EdgeInsets.all(sy(12)),
              color: Colors.blue,
              child: Center(
                child: Text(
                  "Scaled Text 3",
                  style: TextStyle(
                    fontSize: sy(24),
                    color: Colors.white,
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/relative_scale_base.dart
import 'package:flutter/material.dart';

/// DEPRECATED - Use `RelativeBuilder` instead.
/// See example below:
/// ```dart
/// @override
/// Widget build(BuildContext context) {
///   return RelativeBuilder(
///     builder: (context, height, width, sy, sx) {
///       return Container(
///         height: height,
///         width: width,
///         child: Center(
///           child: Column(
///             mainAxisSize: MainAxisSize.min,
///             children: [
///               SizedBox(
///                 height: sy(30),
///                 width: sy(30),
///               ),
///               SizedBox(
///                 height: sx(30),
///                 width: sx(30),
///               ),
///             ],
///           ),
///         ),
///       );
///     },
///   );
/// }
/// ```
@Deprecated(
    "This mixin is no longer recommended. Use \"RelativeBuilder\" instead.")
mixin RelativeScale {
  double? _screenHeight;
  double? _screenWidth;

  /// `RelativeScale.screenHeight` -> the height of the screen.
  double? get screenHeight => _screenHeight;

  /// `RelativeScale.screenWidth` -> the width of the screen.
  double? get screenWidth => _screenWidth;

  /// Initialize `RelativeScale`.
  initRelativeScaler(BuildContext context) {
    var size = MediaQuery.of(context).size;
    _screenHeight = size.height;
    _screenWidth = size.width;
  }

  /// Size relative to the `height` of the screen.
  double sy(double value) {
    return (_screenHeight! * _calculate(value)).roundToDouble();
  }

  /// Size relative to the `width` of the screen.
  double sx(double value) {
    return (_screenWidth! * _calculate(value)).roundToDouble();
  }

  double _calculate(double value) {
    return (value / 100) / 5.333333333333333;
  }
}

typedef _RelativeBuilderFunction = Widget Function(
  BuildContext,
  double,
  double,
  double Function(double),
  double Function(double),
);

/// A relative builder specially made for stateless widget which can't
/// inherit from the mixin RelativeScale.
class RelativeBuilder extends StatelessWidget {
  /// Scale the sizes the way you want. Defaults to `1.0`.
  /// If `RelativeBuilder` turns your sizes a bit bigger or smaller
  /// you can decrease or increase the scale to match your expected sizes.
  final double? scale;

  /// ## Parameters
  /// `context` - The [BuildContext] for this builder widget.
  ///
  /// `height` - The height of the screen.
  ///
  /// `width` - The width of the screen.
  ///
  /// `sy(double)` - A function which scales the value specified
  /// based on screen height.
  ///
  /// `sx(double)` - A function which scales the value specified
  /// based on screen width.
  ///
  /// ## Example
  /// ```dart
  /// builder: (context, height, width, sy, sx) {
  ///   return Container(
  ///     height: height,
  ///     width: width,
  ///     child: Text(
  ///       'Scaled Text!!',
  ///        style: TextStyle(fontSize: sy(12)),
  ///     ),
  ///   );
  /// }
  /// ```
  final _RelativeBuilderFunction builder;

  /// A relative builder specially made for stateless widget which can't
  /// inherit from the mixin RelativeScale.
  const RelativeBuilder({
    Key? key,
    this.scale,
    required this.builder,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return _RelativeBuilder(scale: scale, builder: builder);
  }
}

class _RelativeBuilder extends StatefulWidget {
  final double? scale;
  final _RelativeBuilderFunction? builder;

  const _RelativeBuilder({
    Key? key,
    this.scale,
    this.builder,
  }) : super(key: key);

  @override
  _RelativeBuilderState createState() => _RelativeBuilderState();
}

class _RelativeBuilderState extends State<_RelativeBuilder> {
  @override
  Widget build(BuildContext context) {
    return widget.builder!(
      context,
      _screenHeight,
      _screenWidth,
      _sy,
      _sx,
    );
  }

  Size get _size {
    return MediaQuery.of(context).size;
  }

  double get _screenHeight => _size.height;
  double get _screenWidth => _size.width;

  double _sy(double value) {
    return (_screenHeight * _calculate(value)).roundToDouble();
  }

  double _sx(double value) {
    return (_screenWidth * _calculate(value)).roundToDouble();
  }

  double _calculate(double value) {
    return ((value / 100) / 5.333333333333333) * (widget.scale ?? 1);
  }
}


example/example.dart
import 'package:flutter/material.dart';
import 'package:relative_scale/relative_scale.dart';

class ScaledExample extends StatefulWidget {
  ScaledExample({Key key}) : super(key: key);

  @override
  _ScaledExampleState createState() => _ScaledExampleState();
}

// RelativeScaler is a mixin
class _ScaledExampleState extends State<ScaledExample> with RelativeScale {
  @override
  Widget build(BuildContext context) {
    // this is very important, you should always call this whenever you implement RelativeScaler on a widget.
    initRelativeScaler(context);

    /* You'll see "sy" and "sx" below: */
    /* which means "sy" = size relative to height of the screen. */
    /* which means "sx" = size relative to width of the screen. */

    return Container(
      height: screenHeight,
      width: screenWidth,
      child: SingleChildScrollView(
        child: Column(
          children: <Widget>[
            Container(
              height: sy(160),
              width: sy(240),
              margin: EdgeInsets.all(sy(12)),
              color: Colors.redAccent,
              child: Center(
                child: Text(
                  "Scaled Text 1",
                  style: TextStyle(
                    fontSize: sy(24),
                    color: Colors.white,
                  ),
                ),
              ),
            ),
            Container(
              height: sy(80),
              width: sy(280),
              margin: EdgeInsets.all(sy(12)),
              color: Colors.green,
              child: Center(
                child: Text(
                  "Scaled Text 2",
                  style: TextStyle(
                    fontSize: sy(24),
                    color: Colors.white,
                  ),
                ),
              ),
            ),
            Container(
              height: sy(120),
              width: sy(240),
              margin: EdgeInsets.all(sy(12)),
              color: Colors.blue,
              child: Center(
                child: Text(
                  "Scaled Text 3",
                  style: TextStyle(
                    fontSize: sy(24),
                    color: Colors.white,
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
